src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncTests.cs

```diff
@@ -8136,18 +8410,304 @@ .locals init (int V_0, //i2
+                        public static T Await<T>(System.Threading.Tasks.ValueTask<T> task) => task.GetAwaiter().GetResult();
+                    }
+                }
+                """;
+
+            var code = """
+                await System.Threading.Tasks.Task<int>.FromResult(1);
+                """;
+
+            var comp = CreateRuntimeAsyncCompilation(code, runtimeAsyncAwaitHelpers: runtimeAsyncAwaitHelpers);
+            comp.VerifyDiagnostics(
+                // (1,7): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T' in the generic type or method 'AsyncHelpers.Await<T>(Task<T>)'
+                // await System.Threading.Tasks.Task<int>.FromResult(1);
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "System.Threading.Tasks.Task<int>.FromResult(1)").WithArguments("System.Runtime.CompilerServices.AsyncHelpers.Await<T>(System.Threading.Tasks.Task<T>)", "T", "int").WithLocation(1, 7)
+            );
+        }
+
+        [Fact]
+        public void ValueTaskTAwaitConstraintsViolation()
+        {
+            var runtimeAsyncAwaitHelpers = """
+                namespace System.Runtime.CompilerServices
+                {
+                    public static class AsyncHelpers
+                    {
+                        public static void AwaitAwaiter<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+                        {}
+                        public static void UnsafeAwaitAwaiter<TAwaiter>(TAwaiter awaiter) where TAwaiter : class, ICriticalNotifyCompletion
+                        {}
+
+                        public static void Await(System.Threading.Tasks.Task task) => task.GetAwaiter().GetResult();
+                        public static void Await(System.Threading.Tasks.ValueTask task) => task.GetAwaiter().GetResult();
+                        public static T Await<T>(System.Threading.Tasks.Task<T> task) => task.GetAwaiter().GetResult();
+                        public static T Await<T>(System.Threading.Tasks.ValueTask<T> task) where T : class => task.GetAwaiter().GetResult();
+                    }
+                }
+                """;
+
+            var code = """
+                await default(System.Threading.Tasks.ValueTask<int>);
+                """;
+
+            var comp = CreateRuntimeAsyncCompilation(code, runtimeAsyncAwaitHelpers: runtimeAsyncAwaitHelpers);
+            comp.VerifyDiagnostics(
+                // (1,7): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T' in the generic type or method 'AsyncHelpers.Await<T>(ValueTask<T>)'
+                // await default(System.Threading.Tasks.ValueTask<int>);
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "default(System.Threading.Tasks.ValueTask<int>)").WithArguments("System.Runtime.CompilerServices.AsyncHelpers.Await<T>(System.Threading.Tasks.ValueTask<T>)", "T", "int").WithLocation(1, 7)
+            );
+        }
```

Comment:
- Consider adding a test for `void`-returning method showing we'll still generate a state machine.
- could be in a later PR: consider adding a test for `await using` and `await foreach` in a runtime-async method